constraints/miden-vm/hash.air

```diff
@@ -0,0 +1,215 @@
+    return cycle_row_6 * binary_not(s[0]') * binary_not(s[1]')
+
+
+### Helper evaluators #############################################################################
+
+# Enforces that column must be binary.
+ev is_binary(main: [a]):
+    enf a^2 = a
+
+
+# Enforces that value in column is copied over to the next row.
+ev is_unchanged(main: [column]):
+    ev column' = column
+
+
+# Enforce selector columns constraints
+ev selector_columns(main: [s[3]]):
+    let f_out = get_f_out(s)
+    let f_out_next = get_f_out_next(s)
+    let f_abp = get_f_abp(s)
+    let f_mpa = get_f_mpa(s)
+    let f_mva = get_f_mva(s)
+    let f_mua = get_f_mua(s)
+
+    # Flag that is true when the performed operation is one of the represented by flags f_abp, 
+    # f_mpa, f_mva or f_mua
+    let f_comp = f_abp + f_mpa + f_mva + f_mua
+
+    # Enforce that selector columns are binary.
+    enf is_binary([selector]) for selector in s
+
+    # Enforce that unless f_out = 1 or f_out' = 1, the values in columns s[1] and s[2] are copied 
+    # over to the nex row.
+    enf is_unchanged([s[1]]) when !f_out & !f_out_next
+    enf is_unchanged([s[2]]) when !f_out & !f_out_next
+
+    # Enforce that if any of f_abp, f_mpa, f_mva, f_mua flags is set to 1, the next value of s[0] 
+    # is 0.
+    enf s[0]' * f_comp = 0
+
+    # Enforce that no invalid combinations of flags are allowed.
+    enf cycle_row_7 * binary_not(s[0]) * s[1] = 0
+
+
+# Enforce node index constraints
+ev node_index(main: [s[3], i]):
+    let f_out = get_f_out(s)
+    let f_mp = get_f_mp(s)
+    let f_mv = get_f_mv(s)
+    let f_mu = get_f_mu(s)
+    let f_mpa = get_f_mpa(s)
+    let f_mva = get_f_mva(s)
+    let f_mua = get_f_mua(s)
+
+    # b is the value of the bit which is discarded during shift by one bit to the right.
+    let b = i - 2 * i'
+
+    # Flag that allows to enforce constraint that b is binary only when a new node is absorbed into
+    # the hasher state (when the hash operation is one of Merkle path verification operations or 
+    # next Merkle path node absorption operations)
+    let f_an = f_mp + f_mv + f_mu + f_mpa + f_mva + f_mua
+
+    # Enforce that b is binary only when a new node is absorbed into the hasher state.
+    enf f_an * (b^2 - b) = 0
+
+    # Enforce that when a computation is finished i = 0.
+    enf f_out * i = 0
+
+    # Enforce that the value in i is copied over to the next row unless we are absorbing a new row 
+    # or the computation is finished.
+    let absorbing_or_comp_finished = 1 - f_an - f_out
+    enf is_unchanged([i]) when absorbing_or_comp_finished
+
+
+# Enforce hasher state constraints
+ev hasher_state(main: [s[3], h[12], i]):
+    let f_mp = get_f_mp(s)
+    let f_mv = get_f_mv(s)
+    let f_mu = get_f_mu(s)
+
+    # Flag that is true when the performed operation includes absorbing the next node during Merkle
+    # path computation.
+    let f_absorb_node = f_mp + f_mv + f_mu
+
+    let f_abp = get_f_abp(s)
```

Comment:
nit: I would prefer to move this up a few rows so it's declared with the other flags at the top of the evaluator